### DISCUSSION TITLE
Support for columnar data in `executemany` (mapping of sequences in addition to sequences of mappings)

### DISCUSSION THREAD

**Discussion by bluenote10** [psycopg/psycopg#428] [General]
So far the `params_seq` argument of `executemany` seems to support either "sequences of sequence" for positional arguments and "sequences of mappings" for named arguments.

This means that if a query touches a large number of rows, one has to create a huge number of temporary dicts, one for each row, basically with the same/redundant dictionary keys for each record. This has a significant overhead in terms of heap allocations, and is tedious in the cases where the user data is in columnar form.

It would be nice if the `params_seq` would allow for passing "mappings of sequences" as well:

```py
# sequence of mappings
cur.executemany(query_template, [
    {"foo": 1, "bar": 42},
    {"foo": 2, "bar": 43},
])

# mappings of sequences
cur.executemany(query_template, [
    {"foo": [1, 2]},
    {"bar": [42, 43]},
])
```

This would be particularly helpful when working with data frames or numpy arrays, i.e., could play nicely with https://github.com/psycopg/psycopg/issues/192 ?

**Comment by dvarrazzo** [comment]
If Postgres can understand something, then psycopg should be able to send it. The point here is: what is `query_template`? Are we able to ask Postgres a column-oriented insert or update?

With one column, you can use unnest. With more than one column you would need some sort of transpose, or parallel unnest. Does Postgres provide this functionality?

**Reply by bluenote10** [reply]
I'm not a Postgres expert, but most likely Postgres doesn't have support for columnar insert/update.

I didn't really mean to send the data in columnar for to Postgres and unnest there, but rather just to internally transform the second form into the first form within psycopg. I.e., this was mainly just a feature suggestion that the "transpose" could be done internally in psycopg. This would be nice from a user perspective (because having to do such kind of transposes is now often needed on use site) and perhaps it can be done more efficiently, because it may not be necessary to create all the temporary dicts in memory at once.

**Reply by dvarrazzo** [reply]
You don't have to create all the temporary dicts either: you can pass excutemany a generator and, create one dict at time.
